activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -107,7 +107,12 @@ def select_all(arel, name = nil, binds = [], preparable: nil, async: false) # :n
 
           if async
             result = lookup_sql_cache(sql, name, binds) || super(sql, name, binds, preparable: preparable, async: async)
-            FutureResult::Complete.new(result)
+
+            if result.is_a?(FutureResult) || result.is_a?(FutureResult::Complete)
```

Comment:
We need to test for both classes, because when async is configured in the app - we have `FutureResult::SelectAll` class and when not, `FutureResult::Complete`.

See https://github.com/rails/rails/blob/25b7f640749e0dbbf731a7f5696e5c785195f922/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L609-L636

Reply:
Good spot!

I don't suppose we can find a solution involving duck typing or something different more polymorphy-friendly? 🤔 

At least we could move this `if / else` in a method within the `FutureResult` module, to make it more likely for this case to be added if another class is added there?

```ruby
FutureResult.wrap(result)
```

Reply:
Yes. Would be good to introduce a predicate that is common to all `Result` classes. e.g. `Result#future?`

Reply:
But `FutureResult.wrap` is nice too.

Reply:
Thank you for suggestions! ❤️ I implemented a `FutureResult.wrap` and also update a few places in the codebase to use it.